### ISSUE TITLE
Feature Request: Pass Stan model as string to CmdStanModel

### ISSUE THREAD

**Issue by fonnesbeck** [stan-dev/cmdstanpy#757]
#### Summary:

Add an optional argument for passing Stan model as string to `CmdStanModel`

#### Description:

It would be a huge convenience to be able to pass a Stan model to `CmdStanModel` as a string, rather than first having to write it to a file. Adding boilerplate code to write a string to a file from Python only to load it back for the model to use seems unnecessary from a user's standpoint.

#### Current Version:
1.2.2

**Comment by WardBrian** [comment]
This was last discussed here: 
https://github.com/stan-dev/cmdstanpy/issues/513

The discussion at the time mainly focused on how Stan code is code, so it makes sense to live in its own files that can be separately version controlled etc.

I think the way we would implement this would also force you to re-compile each time Python got restarted, which is also downside to allowing this

**Comment by vankesteren** [comment]
Just to add to this point, feel free to ignore or explicitly choose `wontfix` (I appreciate clear scope decisions in a project like this). 

I understand that you want to instil good coding practice into your user base. But having this as a feature would be really nice nonetheless for quick experimentation. I miss it when I want to explain or try out some small model. For new users, the barrier will be lowered if you allow stan code to be passed. It also has advantages for developers: a minimal reprex will be much easier to create and read, for example in github issues. Compare these:

### Option 1: current situation
```py
from cmdstanpy import CmdStanModel
from tempfile import TemporaryFile

stan_string = """
data {
  int<lower=0> N;
  array[N] int<lower=0,upper=1> y;
}
parameters {
  real<lower=0,upper=1> theta;
}
model {
  theta ~ beta(1,1);
  y ~ bernoulli(theta);
}
"""

with TemporaryFile(mode="w", suffix=".stan", delete=False) as file:
    file.writelines(stan_string)

stan_model = CmdStanModel(stan_file=file.name)
stan_data = {"N": 10, "y": [0, 1, 0, 0, 0, 0, 0, 0, 0, 1]}
stan_fit = stan_model.sample(stan_data)
```

### Option 2: allowing stan code as a string in the constructor
```py
from cmdstanpy import CmdStanModel

stan_string = """
data {
  int<lower=0> N;
  array[N] int<lower=0,upper=1> y;
}
parameters {
  real<lower=0,upper=1> theta;
}
model {
  theta ~ beta(1,1);
  y ~ bernoulli(theta);
}
"""
stan_model = CmdStanModel(stan_code=stan_string)
stan_data = {"N": 10, "y": [0, 1, 0, 0, 0, 0, 0, 0, 0, 1]}
stan_fit = stan_model.sample(stan_data)
```

The second is not only smaller, but also much easier to read for newcomers. File I/O is quite verbose in python (I did my best to keep it small in example 1) and it might distract from the actual content.

An alternative option is to make a very explicit `from_string` classmethod:

### Option 3: `from_string` classmethod
```py
from cmdstanpy import CmdStanModel

stan_string = """
data {
  int<lower=0> N;
  array[N] int<lower=0,upper=1> y;
}
parameters {
  real<lower=0,upper=1> theta;
}
model {
  theta ~ beta(1,1);
  y ~ bernoulli(theta);
}
"""
stan_model = CmdStanModel.from_string(stan_string)
stan_data = {"N": 10, "y": [0, 1, 0, 0, 0, 0, 0, 0, 0, 1]}
stan_fit = stan_model.sample(stan_data)
```

As I'm writing this, I think this would be quite nice actually. It separates the code implementation clearly, it makes it explicit that this is not the "standard way" of doing things, and could even warn users about bad coding practices.

Whether this accessibility / coding practice tradeoff is worth it is entirely up to you to decide! 

I checked, and `cmdstanr` also does not support this directly but it does expose a function `write_stan_file`. That would make the reprex like this:

### Option 4: A write_stan_file function
```py
from cmdstanpy import CmdStanModel, write_stan_file

stan_string = """
data {
  int<lower=0> N;
  array[N] int<lower=0,upper=1> y;
}
parameters {
  real<lower=0,upper=1> theta;
}
model {
  theta ~ beta(1,1);
  y ~ bernoulli(theta);
}
"""
stan_path = write_stan_file(stan_string)
stan_model = CmdStanModel(stan_file=stan_path)
stan_data = {"N": 10, "y": [0, 1, 0, 0, 0, 0, 0, 0, 0, 1]}
stan_fit = stan_model.sample(stan_data)
```

This seems a bit un-pythonic and was rejected as an idea in #513.

**Comment by amas0** [comment]
Thanks for writing this to clearly and going through options! 

I can see the argument on both sides for this. And especially with a stan language server now available, having a separate file does just make the process of writing stan code nicer. However with quick snippets, having to maintain a separate file is a bit of friction. 

If we were to write to a temporary file like this, I'd want to have a think if there's a more natural place to put it. We already create temporary directories for the cmdstan output. It may make sense to have temporary Stan files written to the same existing tempdir directory (I haven't looked at the linked PR to see if this does that).

Defer to @WardBrian on whether this is something we want to pursue for inclusion in 2.0. I agree it can lead to worse coding practices, but it's come up enough that it might be worth adding.

**Comment by cdc-mitzimorris** [comment]
have you tried Stan playground for quick experimentation?
https://github.com/flatironinstitute/stan-playground

**Comment by WardBrian** [comment]
I am the last of the active cmdstanpy developers to weigh in between this and #849, which I guess means it comes down to me in some sense. 

I think there are lots of good reasons to *not* accept the code as a string.

1. (the sort of ur-reason from which all the rest derive) CmdStan is a file-based interface with a file-based build system. At some point a file _must_ exist, and I think it is better if the user directly decides where that file is.
2. Most basic solutions (like #849) break the nice feature that a model will not be re-compiled unless the code changes. Temporary files will be re-built every time. 
  More complicated solutions, like ones that hash the contents to try to produce a _predictable_ filename in order to get cache-hits from the build process, have also lead to subtle bugs in practice (seen in the R interfaces that do this), and require that the 'temporary' files not actually be cleaned up by cmdstanpy, but instead rely on e.g. the OS regularly cleaning up `/tmp` to avoid polluting it over time.
3. "Stan code is code"-like reasons, e.g. if you have it as its own file then you can gain access to syntax highlighting, linting, formatting, separate version control of that file, etc, etc.
4. If something goes wrong with cmdstanpy, one of the first things I try and suggest to users is to run the model on the command line (_outside_ Python), which becomes harder if the code was written in code-in-strings style. 

Of course, adding it as an option doesn't remove these benefits from people who still write the files separately, but I do think that default behavior matters and that it is worth pushing users, new and old, toward a more idiomatic style through our API design.

Simultaneously, I am unconvinced by the arguments in favor. 

A big one, as mentioned by @amas0 in https://github.com/stan-dev/cmdstanpy/pull/849#issuecomment-3812973817, is notebook-like environments where a single source file really is 'the way'. But I think notebooks are worth their own special consideration, which is why I've built out https://github.com/WardBrian/CmdStanJupyter. 
In terms of the code that is run it is not much more than something akin to the functionality suggested above (storing it in a directory alongside the notebook rather than a temporary one), but it provides syntax highlighting and keeps the Stan code in its own special cell rather than mixed about with the Python.

Outside of notebooks, if someone really does just want temp-file behavior, I don't think it's that bad to write what @vankesteren did above:
```python
with TemporaryFile(mode="w", suffix=".stan", delete=False) as file:
    file.writelines(stan_string)

stan_model = CmdStanModel(stan_file=file.name)
```

If they found themselves doing this a lot (though the primary argument in favor seems to be convenience, so I'm not sure why they would be doing it a lot?), they'd be free to encapsulate it in a function in their own code

---

In summary, my vote is `wontfix`

**Comment by bob-carpenter** [comment]
We usually work by consensus and I'm happy to go along with @WardBrian and @mitzimorris on this one.  We only vote when our developers fundamentally disagree.  So I'm closing.

P.S.  As far as recompiles go, I've noticed that on my new M3-Ultra Mac Studio, the compile time for Stan models is on the order of 2 seconds.  Way faster than anything else I've ever used.
